### PR TITLE
[LED-138] View and Edit notes in table

### PR DESCRIPTION
**Pull request by xxshadowfang**
I should really add a description here so people have an idea of what's going on

### CODE REVIEW

File: src/main/java/ledger/user_interface/ui_controllers/NoteEditInput.java [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+import ledger.controller.register.TaskWithArgs;
+import ledger.database.entity.Note;
+
+import java.net.URL;
+import java.util.ResourceBundle;
+
+/**
+ * Created by gert on 1/25/17.
+ */
+public class NoteEditInput extends GridPane implements IUIController, Initializable {
```
Comment (shellajt) [inline src/main/java/ledger/user_interface/ui_controllers/NoteEditInput.java]: Needs a JavaDoc

File: src/main/java/ledger/user_interface/ui_controllers/NoteEditInput.java [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+    private static final String pageLoc = "/fxml_files/NoteEditInput.fxml";
+    @FXML
+    private TextArea noteText;
+    @FXML
+    private Button saveButton;
+    private Note note;
+
+    public NoteEditInput(){
```
Comment (shellajt) [inline src/main/java/ledger/user_interface/ui_controllers/NoteEditInput.java]: JavaDoc

File: src/main/java/ledger/user_interface/ui_controllers/NoteEditInput.java [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+    }
+    @Override
+    public void initialize(URL location, ResourceBundle resources) {
+        saveButton.setOnAction((event -> {
+            note.setNoteText(noteText.getText());
+            TaskWithArgs<Note> updateNoteTask = DbController.INSTANCE.editNote(note);
+            updateNoteTask.startTask();
+        }));
+    }
+    public void setNote(Note note){
```
Comment (shellajt) [inline src/main/java/ledger/user_interface/ui_controllers/NoteEditInput.java]: JavaDoc

File: src/main/java/ledger/user_interface/ui_controllers/NoteEditInput.java (hunk 1 above)
Comment (shellajt) [inline src/main/java/ledger/user_interface/ui_controllers/NoteEditInput.java]: Lets also use the naming scheme we have for controllers

File: src/main/java/ledger/user_interface/ui_controllers/TransactionTableView.java [hunk 4]
```diff
@@ -392,4 +408,23 @@ public void updateSearchFilterString(String searchFilterString) {
         this.searchFilterString = searchFilterString;
         this.asyncTableUpdate();
     }
+
+    public class MyCustomToggleCell<TransactionModel> extends TableCell<TransactionModel, Boolean> {
```
Comment (shellajt) [inline src/main/java/ledger/user_interface/ui_controllers/TransactionTableView.java]: Lets document this inner class. A meaningful name would be good too

File: src/main/java/ledger/user_interface/ui_controllers/TransactionTableView.java [hunk 5]
```diff
@@ -392,4 +408,23 @@ public void updateSearchFilterString(String searchFilterString) {
...
+        private Button button = new Button();
+
+        public MyCustomToggleCell(TableRowExpanderColumn<TransactionModel> column) {
```
Comment (shellajt) [inline src/main/java/ledger/user_interface/ui_controllers/TransactionTableView.java]: Javadocs on both methods

File: src/main/java/ledger/user_interface/ui_controllers/TransactionTableView.java [hunk 6]
```diff
@@ -14,12 +14,15 @@
 import javafx.scene.control.cell.TextFieldTableCell;
 import javafx.scene.input.KeyCode;
 import javafx.scene.input.KeyEvent;
+import javafx.scene.layout.HBox;
```
Comment (shellajt) [inline src/main/java/ledger/user_interface/ui_controllers/TransactionTableView.java]: You dont use these imports

**Comment by xxshadowfang** [comment]
I can fix the Null issue, I assumed we always had a note instantiated even if it was empty. Ill add Javadocs, and I can make it collapse

With regard to the textarea being in only the note column, I thought the whole point of view was so you weren't working in a cramped space.

File: src/main/java/ledger/user_interface/ui_controllers/TransactionTableView.java [hunk 7]
```diff
@@ -220,6 +223,17 @@ private void asyncTableUpdate() {
     }
 
     private void configureTransactionTableView() {
+        TableRowExpanderColumn<TransactionModel> expanderColumn = new TableRowExpanderColumn<TransactionModel>(param->{
+
+            Note note = param.getValue().getTransaction().getNote();
```
Comment (cjm721) [inline src/main/java/ledger/user_interface/ui_controllers/TransactionTableView.java]: This note object is never used so what is it for?

**Comment by LadyMopar** [comment]
Okay so now if I type in a new note and click save, then open up the view again, my note doesn't show up. Shouldn't it be saved and show up again?

**Comment by xxshadowfang** [comment]
Alright That should have fixed it.

The issue was from when a transaction was created with a null note. I handled the null but I didnt handle the fact that I needed to call insertNote instead of editNote. 

That took longer to find then I expected.

**Comment by shellajt** [comment]
It'd look a lot cleaner if the box you pop up were to scale with the window. 

image

**Comment by xxshadowfang** [comment]
Alright. Ill get to that in a little. finishing up some check support stuff.

File: src/main/java/ledger/user_interface/ui_controllers/NoteEditInputController.java [hunk 8]
```diff
@@ -0,0 +1,82 @@
...
+    }
+
+    @Override
+    public void initialize(URL location, ResourceBundle resources) {
+
+        saveButton.setOnAction((event -> {
+            note.setNoteText(noteText.getText());
+            TaskWithArgs<Note> updateNoteTask;
+            if(wasNull) {
+                updateNoteTask = DbController.INSTANCE.insertNote(note);
```
Comment (cjm721) [inline src/main/java/ledger/user_interface/ui_controllers/NoteEditInputController.java]: Should use transaction update methods. That way all other things that need to be notified of a transaction update are.

File: src/main/java/ledger/user_interface/ui_controllers/NoteEditInputController.java [hunk 9]
```diff
@@ -0,0 +1,82 @@
...
+            } else
+            {
+                updateNoteTask = DbController.INSTANCE.editNote(note);
+            }
+            updateNoteTask.RegisterSuccessEvent(() -> {
+                collapseMethod.call(false);
+                this.update.call();
```
Comment (cjm721) [inline src/main/java/ledger/user_interface/ui_controllers/NoteEditInputController.java]: Using Transaction DB Calls will also remove the need to call update manually.

File: src/main/java/ledger/user_interface/ui_controllers/TransactionTableView.java [hunk 10]
```diff
@@ -220,6 +223,15 @@ private void asyncTableUpdate() {
     }
 
     private void configureTransactionTableView() {
+        TableRowExpanderColumn<TransactionModel> expanderColumn = new TableRowExpanderColumn<TransactionModel>(param->{
```
Comment (cjm721) [inline src/main/java/ledger/user_interface/ui_controllers/TransactionTableView.java]: Make this its own class. Look how they are in master currently for example.

File: src/main/java/ledger/user_interface/ui_controllers/TransactionTableView.java [hunk 11]
```diff
@@ -392,4 +404,27 @@ public void updateSearchFilterString(String searchFilterString) {
         this.searchFilterString = searchFilterString;
         this.asyncTableUpdate();
     }
+
+    /**
+     * Private Button TableCell that includes the logic for the expand and collapse.
+     * @param <TransactionModel>
+     */
+    private class CollapseExpandButton<TransactionModel> extends TableCell<TransactionModel, Boolean> {
```
Comment (cjm721) [inline src/main/java/ledger/user_interface/ui_controllers/TransactionTableView.java]: This goes into the Column Class (Once Made)

**Comment by taylerhow** [comment]
Any updates on this feature?

**Comment by xxshadowfang** [comment]
That should address all the changes. not sure why That controller is in conflict

File: src/main/java/ledger/user_interface/ui_controllers/component/TransactionTableView.java [hunk 12]
```diff
@@ -152,6 +153,7 @@ private void handleDeleteSelectedTransactionsFromTableView() {
 
     @Override
     public void initialize(URL location, ResourceBundle resources) {
+        this.getColumns().add(NoteColumn.noteColumn());
```
Comment (cjm721) [inline src/main/java/ledger/user_interface/ui_controllers/component/TransactionTableView.java]: Should just add column via the FXML. Should need no code changes in TrasactionTableView to implement feature.

File: src/main/java/ledger/user_interface/ui_controllers/component/tablecolumn/NoteColumn.java [hunk 13]
```diff
@@ -0,0 +1,45 @@
...
+import javafx.scene.control.TableColumn;
+import ledger.database.entity.Transaction;
+import ledger.user_interface.ui_controllers.IUIController;
+import ledger.user_interface.ui_controllers.NoteEditInputController;
+import org.controlsfx.control.table.TableRowExpanderColumn;
+
+/**
+ * Created by gert on 2/7/17.
+ */
+public class NoteColumn extends TableColumn implements IUIController {
```
Comment (cjm721) [inline src/main/java/ledger/user_interface/ui_controllers/component/tablecolumn/NoteColumn.java]: You are not using this class as a TableColumn just as a code block.

File: src/main/java/ledger/user_interface/ui_controllers/NoteEditInputController.java [hunk 14]
```diff
@@ -0,0 +1,74 @@
...
+            });
+            updateNoteTask.startTask();
+            updateNoteTask.waitForComplete();
+        }));
+    }
+
+    /**
+     * @param param this is the TableRowData that will be used to edit the Note data.
+     */
+    public void setTableRowData(TableRowExpanderColumn.TableRowDataFeatures<Transaction> param) {
```
Comment (shellajt) [inline src/main/java/ledger/user_interface/ui_controllers/NoteEditInputController.java]: Lets include a description in the JavaDoc. Function name isnt exactly descriptive, some more info would help

**Comment by xxshadowfang** [comment]
@cjm721 @shellajt  Can you guys check this out now?

**Pull request merged**